presto-raptorx/src/main/java/com/facebook/presto/raptorx/metadata/CommitCleaner.java

```diff
@@ -0,0 +1,233 @@
+import org.jdbi.v3.sqlobject.customizer.Bind;
+import org.jdbi.v3.sqlobject.customizer.BindBean;
+import org.jdbi.v3.sqlobject.statement.SqlBatch;
+
+import javax.inject.Inject;
+
+import java.time.Clock;
+import java.time.Duration;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static com.facebook.presto.raptorx.RaptorErrorCode.RAPTOR_INTERNAL_ERROR;
+import static com.facebook.presto.raptorx.metadata.IndexWriter.chunkIndexTable;
+import static com.facebook.presto.raptorx.metadata.ShardHashing.tableShard;
+import static com.facebook.presto.raptorx.util.DatabaseUtil.createJdbi;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.collect.Iterables.partition;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.toSet;
+
+public class CommitCleaner
+{
+    private static final Duration FAILED_TRANSACTION_AGE = Duration.ofHours(1);
+    private static final int CHUNK_BATCH_SIZE = 1000;
+
+    private final TransactionManager transactionManager;
+    private final MasterCommitCleanerDao masterDao;
+    private final DeletedChunksDao deletedChunksDao;
+    private final List<ShardCommitCleanerDao> shardDao;
+    private final Clock clock;
+
+    @Inject
+    public CommitCleaner(TransactionManager transactionManager, Database database)
+    {
+        this(transactionManager, database, Clock.systemUTC());
+    }
+
+    public CommitCleaner(TransactionManager transactionManager, Database database, Clock clock)
+    {
+        this.transactionManager = requireNonNull(transactionManager, "transactionManager is null");
+
+        this.masterDao = createJdbi(database.getMasterConnection())
+                .onDemand(MasterCommitCleanerDao.class);
+
+        this.deletedChunksDao = createDeletedChunksDao(database);
+
+        this.shardDao = database.getShards().stream()
+                .map(shard -> createJdbi(shard.getConnection()))
+                .map(dbi -> dbi.onDemand(ShardCommitCleanerDao.class))
+                .collect(toImmutableList());
+
+        this.clock = requireNonNull(clock, "clock is null");
+    }
+
+    public synchronized void removeOldCommits()
+    {
+        long activeCommitId = transactionManager.oldestActiveCommitId();
+
+        Set<Long> droppedTableIds = masterDao.getDroppedTableIds(activeCommitId);
+
+        Map<Integer, Set<Long>> droppedTableIdsByShard = droppedTableIds.stream()
+                .collect(groupingBy(tableId -> tableShard(tableId, shardDao.size()), toSet()));
+
+        // drop index tables
+        for (long tableId : droppedTableIds) {
+            shardDao.get(tableShard(tableId, shardDao.size())).useHandle(handle ->
+                    handle.execute("DROP TABLE IF EXISTS " + chunkIndexTable(tableId)));
```

Comment:
what about using IndexWriter.dropIndexTable,  as in DatabaseMetadataWriter line213.

There is no error right now. But usually that's what most codes do.

Reply:
Thanks, I added that method later and forgot to update this place.